Source/WebKit/UIProcess/skia/BackingStoreSkia.cpp

```diff
@@ -46,17 +58,61 @@ BackingStore::~BackingStore() = default;
 
 void BackingStore::paint(PlatformPaintContextPtr cr, const WebCore::IntRect& rect)
 {
-    notImplemented();
+    m_surface->draw(cr, rect.x(), rect.y());
 }
 
 void BackingStore::incorporateUpdate(UpdateInfo&& updateInfo)
 {
-    notImplemented();
+    ASSERT(m_size == updateInfo.viewSize);
+    if (!updateInfo.bitmapHandle)
+        return;
+
+    auto bitmap = WebCore::ShareableBitmap::create(WTFMove(*updateInfo.bitmapHandle));
+    if (!bitmap)
+        return;
+
+#if ASSERT_ENABLED
+    WebCore::IntSize updateSize = updateInfo.updateRectBounds.size();
+    updateSize.scale(m_deviceScaleFactor);
+    ASSERT(bitmap->size() == updateSize);
+#endif
+
+    scroll(updateInfo.scrollRect, updateInfo.scrollOffset);
+
+    // Paint all update rects.
+    WebCore::IntPoint updateRectLocation = updateInfo.updateRectBounds.location();
+
+    SkCanvas* canvas = m_surface->getCanvas();
+    if (!canvas)
+        return;
+
+    WebCore::GraphicsContextSkia graphicsContext(*canvas, WebCore::RenderingMode::Unaccelerated, WebCore::RenderingPurpose::ShareableLocalSnapshot);
+    graphicsContext.setCompositeOperation(WebCore::CompositeOperator::Copy);
+
+    for (const auto& updateRect : updateInfo.updateRects) {
+        WebCore::IntRect srcRect = updateRect;
+        srcRect.move(-updateRectLocation.x(), -updateRectLocation.y());
+        bitmap->paint(graphicsContext, m_deviceScaleFactor, updateRect.location(), srcRect);
+    }
 }
 
 void BackingStore::scroll(const WebCore::IntRect& scrollRect, const WebCore::IntSize& scrollOffset)
 {
-    notImplemented();
+    if (scrollOffset.isZero())
+        return;
+
+    WebCore::IntRect targetRect = scrollRect;
+    targetRect.move(scrollOffset);
+    targetRect.intersect(scrollRect);
+    if (targetRect.isEmpty())
+        return;
+
+    if (!m_scrollSurface)
+        m_scrollSurface = createCairoImageSurfaceWithFastMalloc(m_size, m_deviceScaleFactor);
+
+    //copyRectFromOneSurfaceToAnother(m_surface.get(), m_scrollSurface.get(), scrollOffset, targetRect);
+    //copyRectFromOneSurfaceToAnother(m_scrollSurface.get(), m_surface.get(), { }, targetRect);
```

Comment:
@carlosgcampos any pointers on what should go here? Everything else follows along with the implementation in `BackingStoreCairo.cpp`.